iroh-memesync/src/message.rs

```diff
@@ -0,0 +1,222 @@
+use std::fmt::{Debug, Display};
+
+use bytes::{Bytes, BytesMut};
+use cid::Cid;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use crate::Error;
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum Message {
+    Response(Response),
+    Request(Request),
+}
+
+impl Message {
+    /// Decoding from a given `BytesMut`.
+    pub fn from_bytes(bytes: BytesMut) -> Result<Self, Error> {
+        let msg = bincode::deserialize(&bytes)?;
+        Ok(msg)
+    }
+
+    /// Decode into a `BytesMut` buffer.
+    pub fn into_bytes(self) -> Vec<u8> {
+        let out = bincode::serialize(&self).expect("should always succeed");
+        out
+    }
+
+    pub fn id(&self) -> QueryId {
+        match self {
+            Message::Request(r) => r.id,
+            Message::Response(r) => r.id,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub struct QueryId(u64);
```

Comment:
make it a UUID... you'll be happy you did in the future :P

- some rando browsing this code

Reply:
what would be the practical benefit? this only needs to be unique for the sender-receiver pair, and even if it is not, not much bad happens. uuids are considerably larger though

Reply:
we had uints as graphsync ids originally. 
we eventual made it a UUID, largely to remove a bunch of cruft code matching up IDs with layered protocols like go-data-transfer and filecoin retrieval. Plus I find tracking things with just a single id quite pleasant (as opposed to always appending peer ids)
yes they are larger, it's certainly a tradeoff, maybe not worth making now.
the nice thing about uuids is you can share them between layered or seperate protocols.